flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -296,6 +296,26 @@ public void testReleaseOnSendException() throws Exception {
 		}
 	}
 
+	@Test(timeout = DEFAULT_TEST_TIMEOUT)
+	public void testReleaseOnSendError() throws Exception {
```

Comment:
Just noticed we have now duplicated code in `testReleaseOnSendException/testReleaseOnSendError`,
`createOpenedSendExceptionTestCassandraSink/createOpenedSendErrorTestCassandraSink `
and `SendExceptionTestCassandraSink/SendExceptionTestCassandraSink`.
It differs only by type of thrown `Throwable`.
Could you deduplicate it?

Reply:
I tried a few approaches but was unable to really make it any less duplicated: it all boils down to the fact that even though both `Error` and `Exception` inherits from `Throwable` we are allowed to throw either `Error` and `Exception` without adding to the throws clause --- but we are *not* allowed to throw a `Throwable` without adding `Throwable` to the throws clause of `send()`. Therefore using generics is not possible.

It is possible to use Function, but whether this is more nice or not I will let you decide: I put it in a separate branch https://github.com/apache/flink/compare/master...mchro:generify-cassandra-semaphore-leak-catch-throwable?expand=1

Reply:
This is close to what I had in mind but the test code could still be deduplicated.
I added one more [commit](https://github.com/azagrebin/flink/commit/357dc240469a93b1e3bf0ad8828185aa950731b8) for that with some code style changes.
I cannot push to your PR, it seems to be disabled. You can review my suggestion, include it into your PR if you agree and squash the commits, then I will merge the PR.

Reply:
Thanks, I cherry-picked and squashed into this branch!
I fixed a few formatting errors than `mvn test` provided.

Reply:
ok, thanks, looks good
There is one more commit which actually resolves Throwable checking problem completely:
https://github.com/azagrebin/flink/commit/6f200a33538cfdfd0dbc8fd597d057cfa298f959
Could you also check it and squash into the PR if it is ok? Then I will merge it.

Reply:
Sure!
Personally I would be tempted to keep the `InvalidQueryException`, as this can practically happen, but theoretically testing for `Throwable` should be enough.